### PR TITLE
#744 Group integration tests in an external script

### LINKED ISSUE #744: Unstable build does not complete on develop branch due to GitHub Actions runner timeout
## Checked for duplicates
Yes - I've already checked

## 🐛 Describe the bug
The unstable build CI workflow on the `develop` branch consistently times out. @al-niessner investigated and believes this is a resource overload issue causing the GitHub Actions runner to fail and not complete the build.

## 🕵️ Expected behavior
The unstable build should complete successfully on the `develop` branch without timing out.

## 📜 To Reproduce
1. Push a commit to the `develop` branch
2. Observe the unstable build GitHub Actions workflow
3. Workflow times out before completing

## 🖥 Environment Info
- GitHub Actions runner (hosted)
- Branch: `develop`

## 📚 Version of Software Used
N/A

## 🩺 Test Data / Additional context
Investigated by @al-niessner. Suspected root cause: resource overload on the GitHub Actions runner (CPU/memory/disk pressure).

## 🦄 Related requirements
N/A

---

### LINKED ISSUE #745: Integration tests in unstable build suite do not pass when run locally
## Checked for duplicates
Yes - I've already checked

## 🐛 Describe the bug
The integration tests that are part of the unstable build suite are also failing when run locally. This was discovered by @al-niessner during investigation of the unstable build CI timeout.

## 🕵️ Expected behavior
Integration tests should pass when run locally.

## 📜 To Reproduce
1. Clone the repository
2. Run the integration test suite locally
3. Observe test failures

## 🖥 Environment Info
- Local development environment
- Branch: `develop`

## 📚 Version of Software Used
N/A

## 🩺 Test Data / Additional context
Discovered by @al-niessner during investigation of #744 (unstable build timeout). The local failures indicate the issue is not solely a CI/runner resource problem and may require deeper investigation into the test suite itself.

## 🦄 Related requirements
N/A

---

### PR DESCRIPTION
**Pull request by al-niessner**
## 🗒️ Summary
Pulled the testing from actions to local machine but verify that it has been run and successful in actions. Cheater and inexpensive off loading...

### 🤖 AI Assistance Disclosure
- [ ] No AI assistance used
- [ ] AI used for light assistance (e.g., suggestions, refactoring, documentation help, minor edits)
- [x] AI used for moderate content generation (AI generated some code or logic, but the developer authored or heavily revised the majority)
- [ ] AI generated substantial portions of this code

Estimated % of code influenced by AI: ___ %

## ⚙️ Test Data and/or Report
All tests should pass

## ♻️ Related Issues
Closes #744 
Closes #745 

## 🤓 Reviewer Checklist

*Reviewers: Please verify the following before approving this pull request.*

### Documentation and PR Content
- [ ] **Documentation:** README, Wiki, or inline documentation (Sphinx, Javadoc, Docstrings) have been updated to reflect these changes.
- [ ] **Issue Traceability:** The PR is linked to a valid GitHub Issue
- [ ] **PR Title:** The PR title is "user-friendly" clearly identifying what is being fixed or the new feature being added, that if you saw it in the Release Notes for a tool, you would be able to get the gist of what was done.

### Security & Quality
- [ ] **SonarCloud:** Confirmed no new High or Critical security findings.
- [ ] **Secrets Detection:** Verified that the Secrets Detection scan passed and no sensitive information (keys, tokens, PII) is exposed.
- [ ] **Code Quality:** Code follows organization style guidelines and best practices for the specific language (e.g., PEP 8, Google Java Style).

### Testing & Validation
- [ ] **Test Accuracy:** Verified that test data is accurate, representative of real-world PDS4 scenarios, and sufficient for the logic being tested.
- [ ] **Coverage:** Automated tests cover new logic and edge cases.
- [ ] **Local Verification:** (If applicable) Successfully built and ran the changes in a local or staging environment.

### Maintenance
- [ ] **Backward Compatibility:** Confirmed that these changes do not break existing downstream dependencies or API contracts (or that breaking changes are clearly documented).

### CODE REVIEW

**Comment by al-niessner** [comment]
@jordanpadams @tloubrieu-jpl @nutjob4life 

Y'all are gonna hate me for this. Wrote a script to do the work locally. It saves a JSON file with information to tell github actions if there was success or not. Solves the small action resources even as tests grow. Keeps enforcing testing. Yes, can fake it, but can fake anything. For instance, postman was returning success despite having 12 assertion failures . Turns out the return status is not all that reliable. Have to do some log parsing. Minimal today, but can improve as it evolves.

File: .github/workflows/integration_tests.sh [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+bdir=$(dirname "$(realpath "$0")")
+rdir=$(realpath "$bdir/../..")
+cd "$rdir" || exit 1
+api_gitrev=$(git describe --always --abbrev=40 --dirty='+' --exclude '*')
+tdir=$(mktemp -d)
+# The EXIT pseudo-signal covers normal exits, errors, and interruptions (Ctrl+C)
+trap 'rm -rf "$tdir"' EXIT
+cd "$tdir" || exit 1
+git clone --quiet https://github.com/NASA-PDS/registry.git
```
Comment (nutjob4life) [inline .github/workflows/integration_tests.sh:110]: Is this okay? It looks like we're just cloning HEAD and so makes this repository dependent on whatever the state of `registry` happens to be at the time.

I'm okay with that—just want to double-check.

File: .github/workflows/integration_tests.sh [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+#
+
+build() {
+    mvn --quiet clean package
+    jar_file="$(find ./service/target/ -maxdepth 1 -regextype posix-extended -regex '.*/registry-api-service-[0-9]+\.[0-9]+\.[0-9]+(-SNAPSHOT)?\.jar')"
+    docker build --build-arg api_jar="$jar_file" -t nasapds/registry-api-service:latest -f docker/Dockerfile .
+}
+
+double_check_logfile() {
+    grep -Eq "[[:space:]]*#[[:space:]]+failure[[:space:]]+detail" "$1" \
```
Comment (nutjob4life) [inline .github/workflows/integration_tests.sh:41]: This feels a little brittle. Is it possible to just trust the exit code? 

I could foresee that a change in the tooling (say they change the wording or capitaliation) could break this in a non-obvious or surprising way.

**Review by nutjob4life — CHANGES_REQUESTED** [review]
Hah! Wow, this is … actually pretty nifty, @al-niessner. But I think it could use some docs as well, since the CI itself is now "part of the official record" (git history) with the `last_integration_test.json` file, and there are some new motions developers have to go through:

- README update: explain what `last_integration_test.json`, when it must be updated, how, what GitHub Actions is actually checking
- Maybe this repo should have its own `.github/pull_request_template.md` with an extra checkbox reminding people to do that?

File: .github/workflows/integration_tests.sh (hunk 2 above)
Comment (al-niessner) [reply .github/workflows/integration_tests.sh:41]: Super brittle. The postman output is terrible and only meant to be read by a human. Fancy tables and stuff but it is a pain to code a reader up then match columns and rows. Anyway this is built on a sample size of 1 so expect it to change. I gave its own function because I figured it would be growing and changing as needed.

While hacky and brittle, better than using the erroneous error/status code.

**Comment by al-niessner** [comment]
> Hah! Wow, this is … actually pretty nifty, @al-niessner. But I think it could use some docs as well, since the CI itself is now "part of the official record" (git history) with the `last_integration_test.json` file, and there are some new motions developers have to go through:
> 
> * README update: explain what `last_integration_test.json`, when it must be updated, how, what GitHub Actions is actually checking
> * Maybe this repo should have its own `.github/pull_request_template.md` with an extra checkbox reminding people to do that?

The JSON is regenerated when you run .github/workflows/integration_tests.sh. No manual editing of the JSON file. Question is, do you add a commit hook (no because there is a fixed 8 minute wait in there for reasons I cannot fathom) or something similar. Maybe the push. However, until working robustly and quickly, just do it manually. May need to move the script to make it more obvious - I would suggest a script to call this script because this script pretends it knows where it is.

Oh docs. Sure...

File: .github/workflows/integration_tests.sh (hunk 2 above)
Comment (nutjob4life) [reply .github/workflows/integration_tests.sh:41]: Copy that

**Comment by nutjob4life** [comment]
> No manual editing of the JSON file

Ohh I missed that. Sorry.

Docs for the win!

File: .github/workflows/integration_tests.sh (hunk 1 above)
Comment (al-niessner) [reply .github/workflows/integration_tests.sh:110]: agreed, not good. will update when have to use a registry branch with registry-api build as will have a working example.

**Comment by al-niessner** [comment]
@jordanpadams @nutjob4life @tloubrieu-jpl 

Well, the postman testing is working. What is the deep archive test and is it critical or not? It is going to take me a day to so to add that to the localized I&T testing because it appears to need/want all of the registry stuff up and running.

**Comment by al-niessner** [comment]
@jordanpadams @nutjob4life @tloubrieu-jpl 

Finally, it passes. Merge it and I will start on common error message and clean up any errors I find along the way.

**Review by nutjob4life — APPROVED** [review]
All right, let's see what happens

**Comment by jordanpadams** [comment]
@al-niessner @tloubrieu-jpl we should discuss this before we merge. it looks like @tloubrieu-jpl has made numerous changes to `develop` since this branch was created, so at minimum we should rebase this branch before merging.

**Comment by tloubrieu-jpl** [comment]
Good point, sorry I missed to have a PR for those changes. I need to get use to have Al and others involved in the integration part.
The changes I made are unrelated so the re-base should work well.
 
There is another PR too which works with updated integration tests https://github.com/NASA-PDS/registry-api/pull/753

And I also updated the registry docker compose in the develop branch to make the integration work again, I am not sure if that is used in this PR.

**Comment by al-niessner** [comment]
@jordanpadams @tloubrieu-jpl 

When these changes are run with registry develop branch it fails with errors in log file despite docker saying postman passed. It just means skipping all those tests again. There is a strange works in one environment but not another despite using all the same docker images. The only way to see it, use this script.

**Comment by tloubrieu-jpl** [comment]
Hi @al-niessner ,

The tests in registry/develop branch match with the registry-api in this PR https://github.com/NASA-PDS/registry-api/pull/753.

I guess I should have the test updated in a specific branch with the latest behavior. But that should be ok for now.

Thanks,

File: .github/workflows/branch-cicd.yaml [hunk 3]
```diff
@@ -89,37 +89,42 @@ jobs:
...
-                    git clone --quiet https://github.com/NASA-PDS/registry.git
-                    cd registry/docker/certs
-                    ./generate-certs.sh
-                    cd ..
-                    docker image inspect nasapds/registry-api-service:latest >/dev/null
-                    docker compose \
-                        --ansi never --profile int-registry-batch-loader --project-name registry \
-                        up --quiet-pull --detach
+                  .github/workflows/integration_tests.sh --verify
+                    #git clone --quiet https://github.com/NASA-PDS/registry.git
```
Comment (tloubrieu-jpl) [inline .github/workflows/branch-cicd.yaml]: We should remove the comments or explain why we will still need them.

File: .github/workflows/branch-cicd.yaml [hunk 4]
```diff
@@ -89,37 +89,42 @@ jobs:
...
+                    #    --ansi never --profile int-registry-batch-loader --project-name registry \
+                    #    run --rm --no-TTY reg-api-integration-test-with-wait
+
+#            -
+#                name: Set up Python 3
+#                uses: actions/setup-python@v6
+#                with:
+#                    python-version: '3.13'
+
+#            -
```
Comment (tloubrieu-jpl) [inline .github/workflows/branch-cicd.yaml]: Are we keep this as a dedicted step ? A step above sounds like we should not.

File: .github/workflows/integration_tests.sh [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    python3 -m venv "$tdir"/da
+    # shellcheck disable=SC1091 # cannot find dynamically created script
+    source "$tdir"/da/bin/activate
+    git clone --quiet https://github.com/NASA-PDS/deep-archive.git
+    cd deep-archive || return 1
+    pip install .
+    pds-deep-registry-archive -u http://localhost:8080 -s PDS_ENG urn:nasa:pds:insight_rad::2.1 --debug
+}
+
+double_check_logfile() {
```
Comment (tloubrieu-jpl) [inline .github/workflows/integration_tests.sh:38]: Chatgpt says newman returns a exit code related to the success/failure of the tests. Is that not true ?

Comment (tloubrieu-jpl) [reply .github/workflows/integration_tests.sh:38]: And also I liked to have the test results in the github action logs, although they are lengthy. That would be useful to still see the table summary and the details for the failed tests.

But I don't believe we should spend to much more time on that as we need to move forward with the registry integration and deployment on AWS.

File: .github/workflows/branch-cicd.yaml (hunk 3 above)
Comment (al-niessner) [reply .github/workflows/branch-cicd.yaml]: dead code removed

File: .github/workflows/branch-cicd.yaml (hunk 4 above)
Comment (al-niessner) [reply .github/workflows/branch-cicd.yaml]: dead code removed

File: .github/workflows/integration_tests.sh (hunk 5 above)
Comment (al-niessner) [reply .github/workflows/integration_tests.sh:38]: kept on the computer running the script for real. auto deleted if test is successful.

**Review by tloubrieu-jpl — APPROVED** [review]
Thanks @al-niessner , good to go.
